### ISSUE TITLE
How to switch cpu modes?

### ISSUE THREAD

**Issue by ningzhenyu** [OP-TEE/optee_os#474]
I have tried OP-TEE and its build-in TAs, they can execute almost perfectly in my ubuntu. But as a novice, I still want to learn how to switch cpu modes. I mean, not through APIs defined by OP-TEE, just a directly way.

I know i should leverage to privileged mode by SVC and then use SMC to trigger the mode switch. But how can I achieve this?

To understand the problem, I write some simple and naive codes just like the following:

int main(int argc, char *argv[])
{
	printf("Hello World!\r\n");
	asm volatile(
	"nop\n\t"
	);
	printf("After NOP!\r\n");
	asm volatile(
	"mov r0, r0\n\t"
	);
	printf("After MOV!\r\n");
	asm volatile(
	"svc  0x0\n\t"
	);
	printf("After SVC!\r\n");
	asm volatile(
	"smc  0x0\n\t"
	);
	printf("After SMC!\r\n");
	return 0;
}

And then cross compile it by:
arm-linux-gnueabihf-gcc -mcpu=cortex-a15 -o hello_world2 hello_world.c

Finally, I add the "hello_world2" file to rootfs of OP-TEE, and run OP-TEE through QEMU.

When I tried to execute "hello_world2" in the target linux machine,  NOP and MOV instruction works well, but I got "illegal instruction" when SVC or SMC instruction was executed.

I just want to know how a program working in normal world can change the cpu mode to access secure world and how the secure world then know it comes and can reply this. I also tried to read source code of OP-TEE, but it is really hard for a novice to finish that.

Thanks very much for your help.

**Comment by jenswi-linaro** [comment]
The `svc` instruction does a syscall, but since you haven't initialized any of the registers it's unpredictable which syscall will be executed.

The `smc` instruction can't be executed in user mode.

**Comment by ningzhenyu** [comment]
@jenswi-linaro  Thanks for your reply.
I was trying to use "svc" instruction to enter privileged mode, and then in that mode "smc" instruction can be executed.
I guess you mean if i can initialize the registers to particular values, then "svc" instruction will works and then "smc" instruction also works, is that correct?

**Comment by jenswi-linaro** [comment]
Please note that both `svc` and `smc` triggers exceptions, that is, not only do they switch CPU mode they also change PC to an entry in the `VBAR` or `MVBAR`. Please check the reference manual (available at http://infocenter.arm.com/help/topic/com.arm.doc.ddi0406c/index.html) for more details.

**Comment by ningzhenyu** [comment]
@jenswi-linaro 
Thanks again~
I checked the manual you mentioned, but did not quite understand it....
After reading it, I updated my mind. I guess firstly I should execute a "svc" instruction in normal world. And then add a "svc" handler to handle with the exception cased by the instruction. As the handler will run in un-secure privileged mode, I can execute a "smc" instruction to enter to secure privileged mode. 
And I have to find out how to add a "svc" handler to handle my "svc" exception, I guess this can be mapped from registers. Is this correct?
